### ISSUE TITLE
Mouse freezes, skips, and stutters over the wezterm window, recovers as soon as it leaves.

### ISSUE THREAD

**Issue by ilar** [wez/wezterm#5366]
### What Operating System(s) are you seeing this problem on?

Linux Wayland

### Which Wayland compositor or X11 Window manager(s) are you using?

Ubuntu Mutter

### WezTerm version

20240429-102818-0184e88e

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

Whenever my mouse enters the boundaries of the wezterm window, it immediately either freezes or starts stuttering and I see a 40-70% cpu increase on whatever core wezterm is running on. Wezterm also starts getting very slow- any freezes or stutters the mouse displays also corrosponds to a hiccup in wezterm's responsiveness.

As soon as i stop moving the mouse, or move the mouse outside the exact boundary of the wezterm window, this stops.

### To Reproduce

Open wezterm, move mouse inside the wezterm window.

### Configuration

no config

### Expected Behavior

Mouse to move smoothly and like normal when located over a wezterm window.

### Logs

Debug Overlay
wezterm version: 20240429-102818-0184e88e x86_64-unknown-linux-gnu
Window Environment: Wayland
Lua Version: Lua 5.4
OpenGL: AMD Radeon RX 6800 XT (navi21, LLVM 15.0.7, DRM 3.54, 6.5.0-28-generic) 4.6 (Compatibility Profile) Mesa 23.2.1-1ubuntu3.1
Enter lua statements or expressions and hit Enter.
Press ESC or CTRL-D to exit
>

(no other logs)

**Comment by ilar** [comment]
This must have been something in the OS interacting with with wezterm, because I apt-get update/upgraded and the issue instantly cleared, without even restarting.

Weird.

**Comment by ilar** [comment]
Nevermind, it came back as soon as I locked my computer and unlocked it. Looks like rebooting fixes it until I lock my screen, or if libmesa gets updated by apt.

**Comment by PaulGrandperrin** [comment]
I also have this issue.

NixOS 24.05, Gnome 46.2, Mesa 24.0.7, wezterm 20240203-110809-5046fc22.

It doesn't always happen right when I first start wezterm but it seems that when it happens it doesn't go away.

Whenever the mouse goes over the wezterm window AND moves:
- gnome-shell starts using 100% CPU
- wezterm locks 25% CPU (4cores, 8threads CPU)
- the whole desktop (not just the mouse, everything, including other apps) heavily stutters

The issue doesn't happen when the mouse is not moved, only when it moves, and the stutter seems to get worse with how much and how fast the mouse moves.

**Comment by PaulGrandperrin** [comment]
One important note, I'm using a 1000Hz mouse.

Reducing the polling rate to 125Hz (the minimum) "solves" the issue: Gnome-shell shoots up to 80-90% CPU usage and no stuttering happens.

Setting to 250Hz makes gnome-shell hit 100% and the desktop starts to shutter.

500Hz, more shuttering.

1000Hz, very heavy stuttering.

My screens are 60FPS.

**Comment by maciekglowka** [comment]
Hi, are there any updates on this :) ? I am experiencing the same issue on a fresh Fedora (41) / Gnome install.
I have a hdpi mouse - so that'd confirm the above.

Tested the config for all three possible frontends, with no good result.

**Comment by daniel-j-h** [comment]
For what is worth: I can report having the very same issue, also on a fresh Fedora 41 / Gnome install, with a 1kHz mouse.

With the WebGPU backend in addition to high cpu usage I'm seeing crashes (https://github.com/wez/wezterm/issues/6480) and with the OpenGl or Software renderer I'm seeing gnome-shell spiking in cpu usage.

What seemed to have helped a bit with the sluggishness is setting

    config.window_decorations = 'NONE'

but nothing I have tried seems to fix the gnome-shell high cpu usage when using the mouse and moving it over the wezterm window.

In practice this means
- when I use the 1kHz mouse and move the mouse pointer over the wezterm window gnome-shell uses 60-80% cpu
- when I use the touchpad (<100 Hz) and move the mouse pointer over the wezterm window gnome-shell uses ~5% cpu

---

As a workaround: I went down the rabbit hole trying to reduce the mouse polling rate down from 1kHz as a workaround. If you want to give that a try I recommend reading up starting from here

https://wiki.archlinux.org/title/Mouse_polling_rate

The issues I ran into here
- there's an app called Piper to configure the mouse; this didn't work for me as the mouse was not detected / supported
- I tried hard-coding the mousepoll frequency instead of using the 1kHz as advertised by the mouse with
    `sudo grubby --update-kernel=ALL --args="usbhid.mousepoll=10"`
  unfortunately this also didn't work as there's a kernel bug with usb3 as described in the arch linux wiki in section 6.2
- The only way to reliably figure out the mouse polling rate is using [evhz](https://git.sr.ht/~iank/evhz) because with the previous issue the kernel might tell you it polls now with e.g. 100 Hz but in reality it's still 1kHz

---

All in all I couldn't reduce the mouse polling frequency but I believe the core issue is that a 1kHz mouse shouldn't be an issue in the first place. I'm still unsure if the underlying issue is in wezterm or in gnome-shell / libmutter, tho.

**Comment by maciekglowka** [comment]
Yeah I did some digging as well. Tried to run via cargo and a profiler.

The results are:
- first run - mouse smooth, but no window decorations (wayland issue I guess)
- second run - disabled wayland - window decorations are present, mouse is smooth

I cannot run the regular Wezterm bin without wayland though (I've installed via flatpak, as there are no packages for F41 yet...maybe for a good reason).

**Comment by Aylawoo** [comment]
Came across this now myself, EndeavourOS Linux 6.17.9-arch1-1 / GNOME 49.2/Mutter/Wayland 

Ryzen 5800x3d/RTX 4070 Ti Super/RX 6600/48GB RAM

Shame as Wezterm was the only terminal emulator I ever found that I liked. Moving the mouse over the window slows my whole system to a crawl. Was hoping to find a solution here.

<img width="1332" height="952" alt="Image" src="https://github.com/user-attachments/assets/b8f31a62-027f-46e6-8581-9e433236c646" />
